File: src/tools/pptx_edit.py
#!/usr/bin/env python3
"""Server-side PPTX editor invoked by the microclaw `pptx_edit` tool.

Reads a JSON spec on stdin, performs structured edits using python-pptx,
writes a JSON result to stdout. Stays small and dependency-light on
purpose — the only runtime dep is `python-pptx`, which is already in
the microclaw image.

JSON spec shape:
  {
    "source_path":  "/abs/path/in.pptx",
    "output_path":  "/abs/path/out.pptx",   # omit to inspect only
    "operations":   [ {op, ...}, ... ]      # optional; if absent, just inspect
  }

Result shape:
  {
    "ok": bool,
    "mode": "inspect" | "apply",
    "source": {...slide summary...},
    "operations_applied": int,
    "errors": [str, ...],
    "output_path": str | null
  }
"""
from __future__ import annotations

import json
import sys
import traceback
from copy import deepcopy

try:
    from pptx import Presentation
    from pptx.util import Pt, Emu
    from pptx.dml.color import RGBColor
except ImportError as e:
    print(json.dumps({
        "ok": False,
        "errors": [f"python-pptx not installed: {e}"],
    }))
    sys.exit(2)


# ---------- inspection ----------

def inspect_presentation(prs):
    """Return a JSON-friendly summary of a Presentation."""
    slides = []
    for idx, slide in enumerate(prs.slides):
        layout_name = slide.slide_layout.name if slide.slide_layout else "(unknown)"
        shapes_summary = []
        for s_idx, shape in enumerate(slide.shapes):
            sd = {
                "shape_index": s_idx,
                "name": shape.name,
                "shape_type": str(shape.shape_type),
                "is_placeholder": bool(getattr(shape, "is_placeholder", False)),
                "has_text_frame": bool(getattr(shape, "has_text_frame", False)),
            }
            if getattr(shape, "is_placeholder", False) and shape.placeholder_format is not None:
                sd["placeholder_idx"] = shape.placeholder_format.idx
                sd["placeholder_type"] = str(shape.placeholder_format.type)
            if getattr(shape, "has_text_frame", False):
                paras = []
                for p_idx, para in enumerate(shape.text_frame.paragraphs):
                    paras.append({
                        "paragraph_index": p_idx,
                        "level": para.level,
                        "text": para.text,
                    })
                sd["paragraphs"] = paras
            shapes_summary.append(sd)
        title = ""
        try:
            if slide.shapes.title is not None:
                title = slide.shapes.title.text or ""
        except Exception:
            pass
        slides.append({
            "slide_index": idx,
            "layout_name": layout_name,
            "title": title,
            "shape_count": len(slide.shapes),
            "shapes": shapes_summary,
        })
    return {
        "slide_count": len(prs.slides),
        "slide_width_emu": prs.slide_width,
        "slide_height_emu": prs.slide_height,
        "slides": slides,
    }


# ---------- helpers ----------

def _get_slide(prs, slide_idx):
    if slide_idx < 0 or slide_idx >= len(prs.slides):
        raise IndexError(f"slide {slide_idx} out of range (have {len(prs.slides)} slides)")
    return prs.slides[slide_idx]


def _replace_text_in_text_frame(tf, find, replace, case_sensitive):
    """Replace text run-by-run so formatting is preserved where the
    match is contained inside a single run. Multi-run matches fall
    back to whole-paragraph replacement (loses run-level formatting
    inside that paragraph)."""
    n = 0
    for para in tf.paragraphs:
        for run in para.runs:
            haystack = run.text or ""
            if case_sensitive:
                if find in haystack:
                    run.text = haystack.replace(find, replace)
                    n += haystack.count(find)
            else:
                if find.lower() in haystack.lower():
                    # Case-insensitive but preserve original casing for non-matches.
                    new = []
                    i = 0
                    f_len = len(find)
                    lower = haystack.lower()
                    f_lower = find.lower()
                    while i < len(haystack):
                        if lower[i:i+f_len] == f_lower:
                            new.append(replace)
                            i += f_len
                            n += 1
                        else:
                            new.append(haystack[i])
                            i += 1
                    run.text = "".join(new)
        # Multi-run match fallback: if the joined paragraph text contains the
        # needle but no single run did, replace at the paragraph level.
        joined = para.text or ""
        target = joined if case_sensitive else joined.lower()
        needle = find if case_sensitive else find.lower()
        if needle in target and not any(
            (find in (r.text or "")) if case_sensitive else (find.lower() in (r.text or "").lower())
            for r in para.runs
        ):
            new_text = joined.replace(find, replace) if case_sensitive else \
                       joined.lower().replace(needle, replace)
            # Wipe runs and set text on first run.
            if para.runs:
                para.runs[0].text = new_text
                for r in para.runs[1:]:
                    r.text = ""
            n += joined.count(find) if case_sensitive else target.count(needle)
    return n


# ---------- operations ----------

def op_replace_text(prs, op):
    find = op["find"]
    replace = op.get("replace", "")
    case_sensitive = op.get("case_sensitive", False)
    slide_filter = op.get("slide")  # optional int; None = all
    n = 0
    for s_idx, slide in enumerate(prs.slides):
        if slide_filter is not None and s_idx != slide_filter:
            continue
        for shape in slide.shapes:
            if getattr(shape, "has_text_frame", False):
                n += _replace_text_in_text_frame(shape.text_frame, find, replace, case_sensitive)
    return {"replacements": n}


def op_set_slide_title(prs, op):
    slide = _get_slide(prs, op["slide"])
    title_shape = slide.shapes.title
    if title_shape is None:
        raise ValueError(f"slide {op['slide']} has no title placeholder")
    title_shape.text = op["title"]
    return {"set": True}


def op_set_paragraph_text(prs, op):
    slide = _get_slide(prs, op["slide"])
    shape_idx = op["shape"]
    para_idx = op.get("paragraph", 0)
    new_text = op["text"]
    if shape_idx < 0 or shape_idx >= len(slide.shapes):
        raise IndexError(f"shape {shape_idx} out of range on slide {op['slide']}")
    shape = list(slide.shapes)[shape_idx]
    if not getattr(shape, "has_text_frame", False):
        raise ValueError("shape has no text frame")
    paras = list(shape.text_frame.paragraphs)
    if para_idx < 0 or para_idx >= len(paras):
        raise IndexError(f"paragraph {para_idx} out of range")
    paras[para_idx].text = new_text
    return {"set": True}


def op_add_bullet(prs, op):
    slide = _get_slide(prs, op["slide"])
    shape_idx = op["shape"]
    text = op["text"]
    level = op.get("level", 0)
    if shape_idx < 0 or shape_idx >= len(slide.shapes):
        raise IndexError(f"shape {shape_idx} out of range on slide {op['slide']}")
    shape = list(slide.shapes)[shape_idx]
    if not getattr(shape, "has_text_frame", False):
        raise ValueError("shape has no text frame")
    p = shape.text_frame.add_paragraph()
    p.text = text
    p.level = level
    return {"added": True}


def op_delete_slide(prs, op):
    slide_idx = op["slide"]
    if slide_idx < 0 or slide_idx >= len(prs.slides):
        raise IndexError(f"slide {slide_idx} out of range")
    # python-pptx has no delete_slide, so we manipulate the XML directly.
    xml_slides = prs.slides._sldIdLst  # noqa: SLF001
    slides = list(xml_slides)
    xml_slides.remove(slides[slide_idx])
    return {"deleted": slide_idx}


def op_add_slide(prs, op):
    layout_idx = op.get("layout", 1)
    after_slide = op.get("after_slide")  # None = append; -1 = beginning
    if layout_idx < 0 or layout_idx >= len(prs.slide_layouts):
        raise IndexError(f"layout {layout_idx} out of range (have {len(prs.slide_layouts)})")
    layout = prs.slide_layouts[layout_idx]
    new_slide = prs.slides.add_slide(layout)
    if op.get("title"):
        if new_slide.shapes.title is not None:
            new_slide.shapes.title.text = op["title"]
    body = op.get("body") or []
    if body:
        # Find a body placeholder.
        body_shape = None
        for shape in new_slide.placeholders:
            if shape.placeholder_format.idx != 0 and getattr(shape, "has_text_frame", False):
                body_shape = shape
                break
        if body_shape is not None:
            tf = body_shape.text_frame
            tf.text = body[0] if body else ""
            for line in body[1:]:
                p = tf.add_paragraph()
                p.text = line
    # Reposition if after_slide given.
    if after_slide is not None:
        xml_slides = prs.slides._sldIdLst  # noqa: SLF001
        slides_xml = list(xml_slides)
        last = slides_xml[-1]  # the new one (we just added)
        xml_slides.remove(last)
        target_pos = after_slide + 1
        target_pos = max(0, min(target_pos, len(slides_xml)))
        xml_slides.insert(target_pos, last)
    return {"added": True, "slide_index": len(prs.slides) - 1}


def op_reorder_slides(prs, op):
    new_order = op["order"]
    if sorted(new_order) != list(range(len(prs.slides))):
        raise ValueError(
            f"order must be a permutation of 0..{len(prs.slides) - 1}; got {new_order}"
        )
    xml_slides = prs.slides._sldIdLst  # noqa: SLF001
    slides_xml = list(xml_slides)
    for s in slides_xml:
        xml_slides.remove(s)
    for new_idx in new_order:
        xml_slides.append(slides_xml[new_idx])
    return {"reordered": new_order}


def op_set_font_size(prs, op):
    slide = _get_slide(prs, op["slide"])
    shape_idx = op["shape"]
    size_pt = op["size"]
    para_idx = op.get("paragraph")  # optional; None = all paragraphs
    shape = list(slide.shapes)[shape_idx]
    if not getattr(shape, "has_text_frame", False):
        raise ValueError("shape has no text frame")
    paras = list(shape.text_frame.paragraphs)
    targets = paras if para_idx is None else [paras[para_idx]]
    n = 0
    for p in targets:
        for r in p.runs:
            r.font.size = Pt(size_pt)
            n += 1
    return {"runs_set": n}


def op_set_font_color(prs, op):
    slide = _get_slide(prs, op["slide"])
    shape_idx = op["shape"]
    rgb_hex = op["rgb"].lstrip("#")
    if len(rgb_hex) != 6:
        raise ValueError(f"rgb must be 6 hex chars; got {op['rgb']!r}")
    rgb = RGBColor.from_string(rgb_hex)
    para_idx = op.get("paragraph")
    shape = list(slide.shapes)[shape_idx]
    if not getattr(shape, "has_text_frame", False):
        raise ValueError("shape has no text frame")
    paras = list(shape.text_frame.paragraphs)
    targets = paras if para_idx is None else [paras[para_idx]]
    n = 0
    for p in targets:
        for r in p.runs:
            r.font.color.rgb = rgb
            n += 1
    return {"runs_set": n}


OP_REGISTRY = {
    "replace_text":       op_replace_text,
    "set_slide_title":    op_set_slide_title,
    "set_paragraph_text": op_set_paragraph_text,
    "add_bullet":         op_add_bullet,
    "delete_slide":       op_delete_slide,
    "add_slide":          op_add_slide,
    "reorder_slides":     op_reorder_slides,
    "set_font_size":      op_set_font_size,
    "set_font_color":     op_set_font_color,
}


def main():
    try:
        spec = json.load(sys.stdin)
    except Exception as e:
        print(json.dumps({"ok": False, "errors": [f"invalid JSON spec: {e}"]}))
        sys.exit(1)

    source_path = spec.get("source_path")
    output_path = spec.get("output_path")
    operations = spec.get("operations") or []

    if not source_path:
        print(json.dumps({"ok": False, "errors": ["source_path required"]}))
        sys.exit(1)

    try:
        prs = Presentation(source_path)
    except Exception as e:
        print(json.dumps({
            "ok": False,
            "errors": [f"failed to open {source_path}: {e}"],
        }))
        sys.exit(1)

    source_summary = inspect_presentation(prs)

    # Inspect-only mode: no operations + no output_path → return summary.
    if not operations and not output_path:
        print(json.dumps({
            "ok": True,
            "mode": "inspect",
            "source": source_summary,
            "operations_applied": 0,
            "errors": [],
            "output_path": None,
        }))
        return

    if not output_path:
        print(json.dumps({
            "ok": False,
            "errors": ["output_path required when operations are given"],
        }))
        sys.exit(1)

    # Apply mode.
    errors = []
    applied = 0
    op_results = []
    for i, op in enumerate(operations):
        op_name = op.get("op")
        if op_name not in OP_REGISTRY:
            errors.append(f"op[{i}]: unknown op {op_name!r}")
            continue
        try:
            result = OP_REGISTRY[op_name](prs, op)
            op_results.append({"index": i, "op": op_name, "result": result})
            applied += 1
        except Exception as e:
            errors.append(f"op[{i}] ({op_name}): {e.__class__.__name__}: {e}")
            op_results.append({"index": i, "op": op_name, "error": str(e)})

    try:
        prs.save(output_path)
    except Exception as e:
        print(json.dumps({
            "ok": False,
            "errors": errors + [f"failed to save {output_path}: {e}"],
            "operations_applied": applied,
            "op_results": op_results,
        }))
        sys.exit(1)

    print(json.dumps({
        "ok": True,
        "mode": "apply",
        "source": {
            "slide_count": source_summary["slide_count"],
        },
        "output_path": output_path,
        "operations_applied": applied,
        "op_results": op_results,
        "errors": errors,
    }))


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(json.dumps({
            "ok": False,
            "errors": [f"unexpected error: {e}", traceback.format_exc()],
        }))
        sys.exit(1)

File: src/tools/pptx_edit.rs
//! `pptx_edit` — structured PowerPoint inspection + editing tool.
//!
//! Wraps a small Python helper (`pptx_edit.py`, embedded via `include_str!`)
//! that uses python-pptx to do real edits server-side. The LLM emits a
//! structured operation list as JSON; we shell out to python3 and pipe
//! the spec via stdin. This replaces the old workflow where the agent
//! had to write a multi-line python script via the `bash` tool — that
//! pattern was unreliable on some upstream providers (the model would
//! return an empty completion mid-script-generation).
//!
//! Operations supported by the Python helper (must match `OP_REGISTRY`):
//!   - `replace_text`        — global or per-slide find/replace
//!   - `set_slide_title`     — set the title placeholder text
//!   - `set_paragraph_text`  — set one paragraph in a shape
//!   - `add_bullet`          — append a bullet to a text frame
//!   - `delete_slide`        — remove a slide by index
//!   - `add_slide`           — add a new slide (layout, title, body bullets)
//!   - `reorder_slides`      — reorder by index permutation
//!   - `set_font_size`       — set Pt size on runs
//!   - `set_font_color`      — set RGB hex color on runs
//!
//! Inspect-only mode (no `operations`, no `output_path`) returns the
//! full slide structure so the agent can plan edits without running
//! its own Python.

use async_trait::async_trait;
use serde_json::json;
use std::process::Stdio;
use tokio::io::AsyncWriteExt;
use tokio::process::Command;
use tokio::time::{timeout, Duration};
use tracing::warn;

use microclaw_core::llm_types::ToolDefinition;

use super::{schema_object, Tool, ToolResult};

const PPTX_HELPER_PY: &str = include_str!("pptx_edit.py");
const EXEC_TIMEOUT_SECS: u64 = 60;

pub struct PptxEditTool;

impl PptxEditTool {
    pub fn new() -> Self {
        Self
    }
}

impl Default for PptxEditTool {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Tool for PptxEditTool {
    fn name(&self) -> &str {
        "pptx_edit"
    }

    fn definition(&self) -> ToolDefinition {
        ToolDefinition {
            name: self.name().into(),
            description: r#"Inspect or edit a PowerPoint (.pptx) file using structured operations — preferred over writing python-pptx scripts in bash.

Two modes:

INSPECT (read-only): pass only `source_path`. Returns the deck structure: slide count, per-slide title + layout + shape summary + paragraph text. Use this first to plan edits.

APPLY: pass `source_path`, `output_path`, and a non-empty `operations` array. Each operation is applied in order; the modified deck is saved to `output_path`. Returns per-op success/failure plus a final ok flag.

Operations (give one as `op` in each operation object):
- `replace_text`        : { find, replace, slide?: int, case_sensitive?: bool } — find/replace across all (or one) slide
- `set_slide_title`     : { slide: int, title: str } — set the title placeholder
- `set_paragraph_text`  : { slide: int, shape: int, paragraph: int, text: str } — overwrite one paragraph
- `add_bullet`          : { slide: int, shape: int, text: str, level?: int } — append bullet to text frame
- `delete_slide`        : { slide: int } — remove a slide by index
- `add_slide`           : { layout?: int, after_slide?: int, title?: str, body?: [str] } — append new slide
- `reorder_slides`      : { order: [int] } — full permutation of current indices
- `set_font_size`       : { slide: int, shape: int, paragraph?: int, size: int } — Pt size on runs
- `set_font_color`      : { slide: int, shape: int, paragraph?: int, rgb: "RRGGBB" } — hex color on runs

All slide/shape/paragraph indices are 0-based. If `paragraph` is omitted on font ops, applies to every paragraph in the shape.

Paths must be absolute or relative to the chat workspace (e.g. `attachments/in.pptx` works). Always inspect first to find the right slide/shape/paragraph indices before editing."#.into(),
            input_schema: schema_object(
                json!({
                    "source_path": {
                        "type": "string",
                        "description": "Path to the input .pptx file. Required."
                    },
                    "output_path": {
                        "type": "string",
                        "description": "Where to write the edited .pptx. Required when `operations` is non-empty. Omit for inspect-only."
                    },
                    "operations": {
                        "type": "array",
                        "description": "Ordered list of operations to apply. Empty/absent = inspect only.",
                        "items": {
                            "type": "object",
                            "properties": {
                                "op": {
                                    "type": "string",
                                    "enum": [
                                        "replace_text",
                                        "set_slide_title",
                                        "set_paragraph_text",
                                        "add_bullet",
                                        "delete_slide",
                                        "add_slide",
                                        "reorder_slides",
                                        "set_font_size",
                                        "set_font_color"
                                    ]
                                }
                            },
                            "required": ["op"]
                        }
                    }
                }),
                &["source_path"],
            ),
        }
    }

    async fn execute(&self, input: serde_json::Value) -> ToolResult {
        let source_path = match input.get("source_path").and_then(|v| v.as_str()) {
            Some(s) => s.to_string(),
            None => return ToolResult::error("Missing required parameter: source_path".into()),
        };
        let output_path = input
            .get("output_path")
            .and_then(|v| v.as_str())
            .map(str::to_string);
        let operations = input
            .get("operations")
            .cloned()
            .unwrap_or_else(|| serde_json::Value::Array(vec![]));

        // Build the spec we'll pipe to python via stdin.
        let mut spec = json!({
            "source_path": source_path,
            "operations": operations,
        });
        if let Some(out) = output_path {
            spec["output_path"] = serde_json::Value::String(out);
        }
        let spec_bytes = match serde_json::to_vec(&spec) {
            Ok(b) => b,
            Err(e) => {
                return ToolResult::error(format!("failed to serialise spec: {e}"))
                    .with_error_type("serialize_error");
            }
        };

        // Spawn python3 with the embedded helper passed via `-c` would be
        // unwieldy at this size; pipe the script via stdin alongside a
        // sentinel won't work either. Instead: pass the script via a
        // python3 invocation that reads it from a heredoc-style stdin
        // — but we need stdin for the spec. So we drop the script into
        // a stable cache path on first call (idempotent) and exec it.
        let script_path = match ensure_helper_on_disk() {
            Ok(p) => p,
            Err(e) => {
                return ToolResult::error(format!("failed to stage pptx helper: {e}"))
                    .with_error_type("stage_error");
            }
        };

        let mut child = match Command::new("python3")
            .arg(&script_path)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .spawn()
        {
            Ok(c) => c,
            Err(e) => {
                return ToolResult::error(format!("failed to spawn python3: {e}"))
                    .with_error_type("spawn_error");
            }
        };

        if let Some(mut stdin) = child.stdin.take() {
            if let Err(e) = stdin.write_all(&spec_bytes).await {
                warn!("pptx_edit: failed to write spec to python stdin: {e}");
            }
            // dropping stdin here closes it so python's sys.stdin.read() returns
        }

        let output_fut = child.wait_with_output();
        let output = match timeout(Duration::from_secs(EXEC_TIMEOUT_SECS), output_fut).await {
            Ok(Ok(o)) => o,
            Ok(Err(e)) => {
                return ToolResult::error(format!("python3 wait failed: {e}"))
                    .with_error_type("wait_error");
            }
            Err(_) => {
                return ToolResult::error(format!(
                    "pptx_edit timed out after {EXEC_TIMEOUT_SECS}s"
                ))
                .with_error_type("timeout");
            }
        };

        let stdout = String::from_utf8_lossy(&output.stdout).into_owned();
        let stderr = String::from_utf8_lossy(&output.stderr).into_owned();

        if !output.status.success() {
            // Helper wrote JSON to stdout even on most errors; surface it
            // alongside stderr so the caller sees both.
            let mut msg = stdout;
            if !stderr.trim().is_empty() {
                msg.push_str("\n[stderr]\n");
                msg.push_str(stderr.trim());
            }
            return ToolResult::error(msg)
                .with_error_type("pptx_helper_error")
                .with_status_code(output.status.code().unwrap_or(1));
        }

        // Truncate very long inspect output (large decks dump full paragraph text).
        let mut text = stdout;
        if text.len() > 40_000 {
            text.truncate(40_000);
            text.push_str("\n... (truncated)");
        }
        ToolResult::success(text)
    }
}

/// Stage the embedded Python helper to a stable cache path so successive
/// invocations don't reparse it. We rewrite if missing or stale.
fn ensure_helper_on_disk() -> std::io::Result<std::path::PathBuf> {
    use std::fs;
    use std::io::Write;
    let dir = std::env::temp_dir().join("microclaw-helpers");
    fs::create_dir_all(&dir)?;
    let path = dir.join("pptx_edit.py");
    let needs_write = match fs::read_to_string(&path) {
        Ok(existing) => existing != PPTX_HELPER_PY,
        Err(_) => true,
    };
    if needs_write {
        let mut f = fs::File::create(&path)?;
        f.write_all(PPTX_HELPER_PY.as_bytes())?;
    }
    Ok(path)
}
